src/main/scala/mesosphere/marathon/core/storage/repository/impl/GroupRepositoryImpl.scala

```diff
@@ -0,0 +1,184 @@
+    protected val appRepository: AppRepository)(
+    implicit
+    ir: IdResolver[PathId, StoredGroup, C, K],
+    marshaller: Marshaller[StoredGroup, S],
+    unmarshaller: Unmarshaller[S, StoredGroup],
+    val ctx: ExecutionContext,
+    val mat: Materializer
+) extends GroupRepository {
+  import StoredGroupRepositoryImpl._
+
+  /* TODO: We're bypassing the cache so it won't store StoredGroup, but we should likely cache
+    the root. We should probably have read-after-write consistency...
+    */
+
+  private val storedRepo = {
+    def leafStore(store: PersistenceStore[K, C, S]): PersistenceStore[K, C, S] = store match {
+      case s: BasePersistenceStore[K, C, S] => s
+      case s: LoadTimeCachingPersistenceStore[K, C, S] => leafStore(s.store)
+      case s: LazyCachingPersistenceStore[K, C, S] => leafStore(s.store)
+    }
+    new PersistenceStoreVersionedRepository[PathId, StoredGroup, K, C, S](leafStore(persistenceStore), _.id, _.version)
+  }
+
+  override def root(): Future[Group] = async {
+    val unresolved = await(storedRepo.get(RootId))
+    unresolved.map(_.resolve(this, appRepository)) match {
+      case Some(group) =>
+        await(group)
+      case None =>
+        Group.empty
+    }
+  }
+
+  override def rootVersions(): Source[OffsetDateTime, NotUsed] =
+    storedRepo.versions(RootId)
+
+  override def rootVersion(version: OffsetDateTime): Future[Option[Group]] = async {
+    val unresolved = await(storedRepo.getVersion(RootId, version))
+    unresolved.map(_.resolve(this, appRepository)) match {
+      case Some(group) =>
+        Some(await(group))
+      case None =>
+        None
+    }
+  }
+
+  /**
+    * @inheritdoc
+    * @todo We need to eventually garbage collect as we will accumulate apps/appVersions that neither
+    * the root nor any historical group references.
+    *
+    * When should we run the garbage collection? We should probably not block root updates, but instead have
+    * a kind of "prepareGC"/"commitGC" phase so that a root update will invalidate the GC and start it over.
+    * If in the process of a prepare, storeRoot is called, we should cancel the current prepareGC and start
+    * it again.
+    *
+    * This may be as simple as "get all of the app ids" and compare them to the appIds that are in a group.
+    * If no group has the appId, delete it. We probably don't actually need to delete app versions that are no longer
+    * in use as there is some group still referring to some version of the app, so eventually it will be deleted once
+    * it is no longer in actual usage.
+    *
+    * The garbage collection does not need to run as part of storeRoot (which would likely slow it down by quite
+    * a bit), instead, we should run it out of band (triggered by a storeRoot) and use the prepare/commit/cancellable
+    * described above.
+    */
+  override def storeRoot(group: Group, updatedApps: Seq[AppDefinition], deletedApps: Seq[PathId]): Future[Done] = {
```

Comment:
- a group is stored with references to app+versions
- the app is stored as separate entity

We have to make sure that all app+version references are always available.
We currently limit the number of versions stored per entity hard in BaseRepository.
Perhaps it makes sense to also do this kind of operation in the gc operation.

Please note: the same is true for tasks (reference an app with version).